lib/ansible/modules/system/gconftool2.py

```diff
@@ -197,74 +285,28 @@ def main():
-        if direct and config_source is None:
-            module.fail_json(msg='If "direct" is "yes" then the ' +
-                             '"config_source" must be specified')
-        elif not direct and config_source is not None:
-            module.fail_json(msg='If the "config_source" is specified ' +
-                             'then "direct" must be "yes"')
-
-    # Create a gconf2 preference
-    gconf_pref = GConf2Preference(module, key, value_type,
-                                  value, direct, config_source)
-    # Now we get the current value, if not found don't fail
-    _, current_value = gconf_pref.call("get", fail_onerr=False)
-
-    # Check if the current value equals the value we want to set.  If not, make
-    # a change
-    if current_value != value:
-        # If check mode, we know a change would have occured.
-        if module.check_mode:
-            # So we will set the change to True
-            change = True
-            # And set the new_value to the value that would have been set
-            new_value = value
-        # If not check mode make the change.
-        else:
-            change, new_value = gconf_pref.call(state)
-    # If the value we want to set is the same as the current_value, we will
-    # set the new_value to the current_value for reporting
-    else:
-        new_value = current_value
-
-    facts = dict(gconftool2={'changed': change,
-                             'key': key,
-                             'value_type': value_type,
-                             'new_value': new_value,
-                             'previous_value': current_value,
-                             'playbook_value': module.params['value']})
-
-    module.exit_json(changed=change, ansible_facts=facts)
+    # If direct access was requested, config_source must be provided.
+    if module.params['direct'] and (module.params['config_source'] is None):
+        module.fail_json(msg='If "direct" is "yes", then the "config_source" must be specified')
+
+    # If present state was specified, value and value type must be provided.
+    if module.params['state'] == 'present' and (module.params['value'] is None or module.params['value_type'] is None):
+        module.fail_json(msg='State "present" requires both "value" and "value_type" to be set.')
+
+    # Create wrapper instance.
+    gconftool2 = GConfTool2(module, module.params['direct'], module.params['config_source'], module.check_mode)
+
+    # Process based on different states.
+    if module.params['state'] == 'get':
+        value = gconftool2.get(module.params['key'])
+        module.exit_json(changed=False, value=value)
+    elif module.params['state'] == 'present':
+        changed = gconftool2.set(module.params['key'], module.params['value_type'], module.params['value'])
+        module.exit_json(changed=changed)
+    elif module.params['state'] == 'absent':
+        changed = gconftool2.unset(module.params['key'])
+        module.exit_json(changed=changed)
```

Comment:
This was done on purpose, but I would like to get some feedback (and will test it again) to make sure I understood things correctly (at least initially when I had look at it it made sense, but it was some time ago).

The old implementation does a `module.exit_json(changed=changed, ansible_facts=facts)`, where the `facts` is a dictionary with one key - `gconftool2`. It was my understanding at the time that by doing this you set globally available facts that will be available for subsequent tasks in the same play (unconditionally I might add). I considered that counter-intuitive since I do not think this module fits the category of general-purpose fact reading module. Instead I tried to make the value available for the `get` state via `register` keyword in task definition.

Perhaps to illustrate things a bit better, this is my understanding how the old code worked (keep in mind I might be completely wrong, I just recall testing something like this back then :) ):

```
- name: Read the value of widget font
  gconftool2:
    key: "/desktop/gnome/interface/font_name"
    state: get
# Notice global presence of variable no matter what.
- debug: var=gconftool2
```

In case of new module you would need to be more explicit that you want the value:
```
- name: Read the value of widget font
  gconftool2:
    key: "/desktop/gnome/interface/font_name"
    state: get
  register: fontname
- debug: var=fontname
```